kolena/_experimental/dataset/_evaluation.py

```diff
@@ -184,12 +187,70 @@ def validate_data(left: pd.DataFrame, right: pd.DataFrame) -> None:
         raise IncorrectUsageError("numbers of rows between two dataframe do not match")
 
 
+def _validate_common_columns(common_columns: List[str]) -> None:
+    if len(common_columns) == 0:
+        raise ValueError("there is no common columns")
+
+
+def _get_default_infer_func(df_inf: pd.DataFrame, join_keys: list[str]) -> INFER_FUNC_TYPE:
+    def infer(datapoints: pd.DataFrame) -> pd.DataFrame:
+        inferences = datapoints[join_keys].merge(df_inf, how="left", on=join_keys)
+        return inferences
+
+    return infer
+
+
+def _get_default_eval_func(df_metrics: pd.DataFrame, join_keys: list[str]) -> EVAL_FUNC_TYPE:
```

Comment:
I feel this is too much black magic to understand the default mechanism. Personally I find, if choosing the two, I'd rather do the `infer` myself.

Bringing this approach to discussion, how about
```
def test(dataset, model, infer_df, metrics_df, on=union[func, str]):
```

Reply:
I like the `on` approach. I'm guessing `str` is a column name to merge on, but what is `func`?

Reply:
I mistakenly thought dataframe `merge/join` `on` includes `func` for complex join logic. Any case would be missing with `on` approach through `str, list of str, or array-like`?

Reference

[join](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html)
> on: str, list of str, or array-like, optional
Column or index level name(s) in the caller to join on the index in other, otherwise joins index-on-index. If multiple values given, the other DataFrame must have a MultiIndex. Can pass an array as the join key if it is not already contained in the calling DataFrame. Like an Excel VLOOKUP operation.

[merge](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html)
> on: label or list
Column or index level names to join on. These must be found in both DataFrames. If on is None and not merging on indexes then this defaults to the intersection of the columns in both DataFrames.

Reply:
just updated to expose `on` in the test func

Reply:
> Any case would be missing with on approach through str, list of str, or array-like?
I could see a results df having a `relative_path` column rather than a `locator` column requiring a bit of preprocessing before running `test` but I don't think it's worth worrying about that. 

I think those types have good coverage.